index.ts

```diff
@@ -119,24 +173,40 @@ export async function exportEventsToBigQuery(events: PluginEvent[], { global }:
                 elements = $elements
             }
 
-            return {
-                uuid,
-                event: eventName,
-                properties: JSON.stringify(ingestedProperties || {}),
-                elements: JSON.stringify(elements || {}),
-                set: JSON.stringify($set || {}),
-                set_once: JSON.stringify($set_once || {}),
-                distinct_id,
-                team_id,
-                ip,
-                site_url,
-                timestamp: timestamp ? global.bigQueryClient.timestamp(timestamp) : null,
+            const object: {json: Record<string, any>, insertId?: string} = {
+                json: {
+                    uuid,
+                    event: eventName,
+                    properties: JSON.stringify(ingestedProperties || {}),
+                    elements: JSON.stringify(elements || {}),
+                    set: JSON.stringify($set || {}),
+                    set_once: JSON.stringify($set_once || {}),
+                    distinct_id,
+                    team_id,
+                    ip,
+                    site_url,
+                    timestamp: timestamp,
+                    bq_ingested_timestamp: new Date().toISOString(),
+                }
+            }
+
+            if (global.deduplicateEvents) {
```

Comment:
given we've disabled BQ retrying, is this used for anything?

Reply:
We've disabled retrying via the bigquery library, not via the plugin-server, so yes, this does work as long as the retry doesn't happen more than a few minutes later.

Reply:
I meant why do we need to dedupe?

Reply:
oh, because exceeding timeouts via the async guard doesn't stop the request from succeeding (most times, depending on the case), so we're making two requests with the same data to bigquery, which leads to duplicates.

This helps resolve some of them. (https://cloud.google.com/bigquery/streaming-data-into-bigquery#dataconsistency)

Reply:
I see now, makes sense!